src/Helpers/Tokens/AsymmetricVerifier.php

```diff
@@ -45,11 +46,12 @@ public function __construct(array $jwks)
      */
     protected function checkSignature(Token $token) : bool
     {
-        $tokenKid = $token->getHeader('kid', false);
-        if (! array_key_exists($tokenKid, $this->jwks)) {
+        $tokenKid   = $token->getHeader('kid', false);
+        $signingKey = is_array( $this->jwks ) ? ($this->jwks[$tokenKid] ?? null) : $this->jwks->getKey( $tokenKid );
```

Comment:
Is like the `$this->jwks` variable is used sometimes to store an array of keys, and sometimes as an instance of some object you can call `getKey`. If that's the case, this should be two different variables as their behavior is different.

Reply:
I see it as a "source of JWKS data." I don't love the two types here but I think it's acceptable to allow this to function and to avoid breaking changes. If I could do an overloaded constructor I would but ... PHP.

Reply:
What about some OOP: creating some sort of interface "Jwk Provider" defining a `getKey` method. The `Jwk Fetcher` would implement that sourcing from the cache and the JWKs URI, and you could have a "Hardcoded Jwk Provider" that also implements that but sourcing from an array. That way, you can pass in the constructor an instance of "Jwk Provider" that you know how to consume.

Reply:
Again, we're into breaking change territory, unfortunately. 

I think this implementation is flexible enough to be useful (if you want to go about getting your keys yourself, storing them locally, providing them directly) but not too confusing. I think "provider" is a better name here but we're a bit stuck with what we've got for now.